indexer/indexer.py

```diff
@@ -0,0 +1,337 @@
+        """
+        raise NotImplementedError(
+            'users must define special_fields to use this base class')
+
+    def merge(self, doc_contents, **kwargs):
+        """
+        Merge the document with the contents in ElasticSearch.
+
+        merge() should take the results of get_item() and harmonize it
+        with whatever is present in ElasticSearch to avoid blind overwritting
+        of documents. Users should implement their own protocol.
+
+        :param doc_contents: Current document to be indexed.
+        :return: The harmonized document which can overwrite existing entry.
+        """
+        raise NotImplementedError(
+            'users must define merge to use this base class')
+
+    def load_doc(self, doc_uuid, doc_contents, * args, **kwargs):
+        """
+        Load a document into ElasticSearch.
+
+        Load 'doc_contents' as a document in ElasticSearch (ES) with a
+        uuid 'doc_uuid'.
+
+        :param index_name: The name of the index to load into ElasticSearch.
+        :param doc_type: The name of the document type in ElasticSearch.
+        :param doc_uuid: The uuid which will root the ElasticSearch document.
+        :param doc_contents: The contents that will be loaded into ES.
+        :return:
+        """
+        # Loads the index document file into elasticsearch
+        self.es_client.index(index=self.index_name,
+                             doc_type=self.doc_type,
+                             id=doc_uuid,
+                             body=doc_contents)
+
+    def create_mapping(self, *args, **kwargs):
+        """
+        Create the ElasticSearch mapping.
+
+        To be overwritten by the class inheriting from here.
+        This class should return the dictionary corresponding to the mapping
+        to be loaded into ElasticSearch. It should use the index_mapping_config
+        attribute as appropriate.
+
+        :return: Dictionary describing the ElaticSearch mapping.
+        """
+        raise NotImplementedError(
+            'users must define create_mapping to use this base class')
+
+    def load_mapping(self, *args, **kwargs):
+        """
+        Load the mapping into Elasticsearch.
+
+        This method is responsible for loading the mapping into Elasticsearch.
+        It first creates an index using the instance's attributes, and
+        then loads the mapping by also calling create_mapping(), which
+        will create the object describing the mapping.
+        """
+        # Creates the index
+        self.es_client.indices.create(index=self.index_name,
+                                      body=self.index_settings,
+                                      ignore=[400])
+        # Loads mappings for the index
+        self.es_client.put_mapping(index=self.index_name,
+                                   doc_type=self.doc_type,
+                                   body=self.create_mapping())
+
+
+class FileIndexer(Indexer):
```

Comment:
`FileCentricIndexer` or `FileOrientedIndexer`. `FileIndexer` sounds too much like sth that indexes files.